src/storage-client/src/source/persist_source.rs

```diff
@@ -245,42 +250,8 @@ impl PendingWork {
                 continue;
             }
             match (key, val) {
-                (Ok(SourceData(Ok(row))), Ok(())) => {
-                    if let Some(mfp) = map_filter_project {
-                        let arena = mz_repr::RowArena::new();
-                        let mut datums_local = datum_vec.borrow_with(&row);
-                        for result in mfp.evaluate(
-                            &mut datums_local,
-                            &arena,
-                            time,
-                            diff,
-                            |time| !until.less_equal(time),
-                            row_builder,
-                        ) {
-                            match result {
-                                Ok((row, time, diff)) => {
-                                    // Additional `until` filtering due to temporal filters.
-                                    if !until.less_equal(&time) {
-                                        output.give_at(&self.capability, (Ok(row), time, diff));
-                                        *work += 1;
-                                    }
-                                }
-                                Err((err, time, diff)) => {
-                                    // Additional `until` filtering due to temporal filters.
-                                    if !until.less_equal(&time) {
-                                        output.give_at(&self.capability, (Err(err), time, diff));
-                                        *work += 1;
-                                    }
-                                }
-                            }
-                        }
-                    } else {
-                        output.give_at(&self.capability, (Ok(row), time, diff));
-                        *work += 1;
-                    }
-                }
-                (Ok(SourceData(Err(err))), Ok(())) => {
-                    output.give_at(&self.capability, (Err(err), time, diff));
+                (Ok(SourceData(data)), Ok(())) => {
+                    output.give_at(&self.capability, (data, time, diff));
```

Comment:
This changes the semantics of the fueled operator. Previously, we would yield based on the output of the MFP, now it yields based on its inputs.

Reply:
Yes, that's true, and I should have called it out!

I _think_ the new semantic is closer to the desired behaviour for fuel - proportional to the amount of work done, not necessarily the size of the output. Does that sound right?